# **Setup**

In [18]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import StackingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neural_network import MLPClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import GradientBoostingClassifier
from matplotlib import pyplot as plt
from sklearn.svm import OneClassSVM

from sklearn.metrics import accuracy_score
from sklearn.metrics import roc_auc_score
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import cross_val_score
from sklearn import metrics

from sklearn.metrics import plot_roc_curve
from sklearn.decomposition import PCA
from matplotlib.pyplot import figure
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.preprocessing import StandardScaler


import pandas as pd
import numpy as np

In [19]:
df = pd.read_csv('preprocessed_spam_ham_phishing.csv')

In [20]:
df.columns

Index(['hops', 'missing_subject', 'missing_to', 'missing_content-type',
       'missing_mime-version', 'missing_x-mailer',
       'missing_content-transfer-encoding', 'missing_x-mimeole',
       'missing_x-priority', 'missing_list-id', 'missing_lines',
       'missing_x-virus-scanned', 'missing_status', 'missing_content-length',
       'missing_precedence', 'missing_delivered-to',
       'missing_list-unsubscribe', 'missing_list-subscribe',
       'missing_list-post', 'missing_list-help', 'missing_x-msmail-priority',
       'missing_x-spam-status', 'missing_sender', 'missing_errors-to',
       'missing_x-beenthere', 'missing_list-archive', 'missing_reply-to',
       'missing_x-mailman-version', 'missing_x-miltered', 'missing_x-uuid',
       'missing_x-virus-status', 'missing_x-spam-level',
       'missing_x-spam-checker-version', 'missing_references',
       'missing_in-reply-to', 'missing_user-agent', 'missing_thread-index',
       'missing_cc', 'missing_received-spf', 'missing_x-orig

In [21]:
new_small_feature_list = ['hops', 'missing_subject', 'missing_to', 'missing_content-type',
   'missing_mime-version', 'missing_x-mailer', 'missing_references',
   'missing_content-transfer-encoding', 'missing_x-mimeole',
   'missing_x-priority', 'missing_list-id', 'missing_lines',
   'missing_x-virus-scanned', 'missing_status', 'missing_content-length',
   'missing_precedence', 'missing_delivered-to',
   'missing_list-unsubscribe', 'missing_list-subscribe',
   'missing_list-post', 'missing_list-help', 'missing_x-msmail-priority',
   'missing_x-spam-status', 'missing_sender', 'missing_errors-to',
   'missing_x-beenthere', 'missing_list-archive', 'missing_reply-to',
   'missing_x-mailman-version', 'missing_x-miltered', 'missing_x-uuid',
   'missing_x-virus-status', 'missing_x-spam-level',
   'missing_x-spam-checker-version', 'missing_references',
   'missing_in-reply-to', 'missing_user-agent', 'missing_thread-index',
   'missing_cc', 'missing_received-spf', 'missing_x-original-to',
   'missing_content-disposition', 'missing_mailing-list',
   'missing_x-spam-check-by', 'missing_domainkey-signature',
   'missing_importance', 'missing_x-mailing-list', 'content-encoding-val',
   'received_str_forged', 'str_content-encoding_empty',
   'str_from_question', 'str_from_chevron', 'str_to_chevron',
   'str_to_empty', 'str_message-ID_dollar', 'str_return-path_bounce',
   'str_content-type_texthtml', 'str_precedence_list', 'length_from',
   'num_recipients_to', 'num_recipients_cc', 'number_replies', 'time_zone',
   'x-priority', 'content-length', 'lines', 'day_of_week',
   'date_comp_date_received', 'span_time', 'conseq_num_received_is_one',
   'conseq_received_good', 'conseq_received_bad', 'conseq_received_date',
   'email_match_from_reply-to', 'domain_val_message-id',
   'domain_match_message-id_from', 'domain_match_from_return-path',
   'domain_match_message-id_return-path', 'domain_match_message-id_sender',
   'domain_match_message-id_reply-to', 'domain_match_return-path_reply-to',
   'domain_match_reply-to_to', 'domain_match_to_in-reply-to',
   'domain_match_errors-to_message-id', 'domain_match_errors-to_from',
   'domain_match_errors-to_sender', 'domain_match_errors-to_reply-to',
   'domain_match_sender_from', 'domain_match_references_reply-to',
   'domain_match_references_in-reply-to', 'domain_match_references_to',
   'domain_match_from_reply-to', 'domain_match_to_from',
   'domain_match_to_message-id', 'domain_match_to_received', 'label']

new_small_feature_list = ['missing_x-mailer', 'missing_reply-to', 'missing_in-reply-to',
   'hops', 'content-encoding-val', 'missing_errors-to', 'missing_sender',
   'received_str_forged', 'str_content-encoding_empty', 'missing_content-transfer-encoding',
   'str_from_question', 'str_from_chevron', 'str_to_chevron',
   'str_to_empty', 'str_message-ID_dollar', 'str_return-path_bounce',
   'str_content-type_texthtml', 'str_precedence_list', 'length_from',
   'num_recipients_to', 'num_recipients_cc', 'number_replies',
   'x-priority',
   'date_comp_date_received', 'span_time', 'conseq_num_received_is_one',
   'conseq_received_good', 'conseq_received_bad', 'conseq_received_date',
   'email_match_from_reply-to', 'domain_val_message-id',
   'domain_match_message-id_from', 'domain_match_from_return-path',
   'domain_match_message-id_return-path', 'domain_match_message-id_sender',
   'domain_match_message-id_reply-to', 'domain_match_return-path_reply-to',
   'domain_match_reply-to_to', 'domain_match_to_in-reply-to',
   'domain_match_errors-to_message-id', 'domain_match_errors-to_from',
   'domain_match_errors-to_sender', 'domain_match_errors-to_reply-to',
   'domain_match_sender_from', 'domain_match_references_reply-to',
   'domain_match_references_in-reply-to', 'domain_match_references_to',
   'domain_match_from_reply-to', 'domain_match_to_from',
   'domain_match_to_message-id', 'domain_match_to_received', 'label']

df = df[new_small_feature_list]

**Remove spam emails, only consider ham and phishing:**

In [22]:
df = df[df['label'] != 1]
print(df.shape)

(26508, 52)


# **Testing:**

**Novelty detection (One Class SVM):**

**Creating a training set composed of only ham emails (10,000 for the training set, rest for testing):**

In [23]:
df_ham = df[df['label'] == 0]
df_phish = df[df['label'] == 2]
print(df_ham.shape)
print(df_phish.shape)

(25220, 52)
(1288, 52)


In [24]:
df_ham_train = df_ham.iloc[0:20000]
df_ham_test = df_ham.iloc[20000:21288]

df_ham_test_Y = df_ham_test['label']
df_ham_train_Y = df_ham_train['label']

df_ham_train_X = df_ham_train.drop('label', axis=1)
df_ham_test_X = df_ham_test.drop('label', axis=1)

feature_list = df_ham_train_X.columns

print(df_ham_test_Y.shape)
print(df_ham_train_X.shape)
print(df_ham_test_X.shape)

(1288,)
(20000, 51)
(1288, 51)


In [25]:
df_phish_test = df_phish.iloc[0:5220]

df_phish_test_Y = df_phish_test['label']
df_phish_test_X = df_phish_test.drop('label', axis=1)

print(df_phish_test_Y.shape)
print(df_phish_test_X.shape)

(1288,)
(1288, 51)


Training just uses ham. Testing uses 1288 ham and 1288 phishing emails. Combing the ham and phishing emails to make the test set:

In [26]:
df_test_X = pd.concat([df_phish_test_X, df_ham_test_X], ignore_index=True)
df_test_Y = pd.concat([df_phish_test_Y, df_ham_test_Y], ignore_index=True)

In [27]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

scaler.fit(df_test_X)
df_test_X = scaler.transform(df_test_X)
df_test_X = pd.DataFrame(df_test_X, columns=feature_list)

scaler = StandardScaler()

scaler.fit(df_ham_train_X)
df_ham_train_X = scaler.transform(df_ham_train_X)
df_ham_train_X = pd.DataFrame(df_ham_train_X, columns=feature_list)

In [28]:
pca = PCA(n_components=20)

#train_X = pca.fit_transform(df_ham_train_X)
#test_X = pca.fit_transform(df_test_X)

In [29]:
test_X = df_test_X
train_X = df_ham_train_X

One class SVM predicts 1 for normal class and -1 for anomaly class, so need to match the labels to those:

In [30]:
df_test_Y = pd.DataFrame(df_test_Y, columns=['label'])

df_test_Y.loc[df_test_Y['label'] == 2, 'label'] = -1
df_test_Y.loc[df_test_Y['label'] == 0, 'label'] = 1


df_train_Y = pd.DataFrame(df_ham_train_Y, columns=['label'])

df_train_Y.loc[df_train_Y['label'] == 2, 'label'] = -1
df_train_Y.loc[df_train_Y['label'] == 0, 'label'] = 1

In [31]:
df_test_Y.value_counts()

label
 1       1288
-1       1288
dtype: int64

In [ ]:
from sklearn.inspection import permutation_importance
from matplotlib import pyplot as plt

model = OneClassSVM(gamma=0.01, nu=0.34, kernel='poly', degree=6, tol=0.001, coef0=0, shrinking=True)
model.fit(train_X, df_train_Y)

results = permutation_importance(model, train_X, df_train_Y, 
                                 scoring='accuracy', n_repeats=2, n_jobs=-1)
importance = results.importances_mean

fig = plt.figure(figsize =(25, 7))

importances_sorted = sorted(zip(importance, feature_list), reverse=True)
feature_sorted, importance_sorted = zip(*importances_sorted)

perm_svm_feature_importances = importance_sorted

print("Top features sorted:")
for x, imp in zip(feature_sorted, importance_sorted):
  print('Feature: %s, Score: %f' % (imp, x))

plt.xticks(rotation='vertical')
plt.bar([x for x in importance_sorted], feature_sorted, width=0.3)
plt.show()

In [ ]:
counter = 0
for x, imp in zip(feature_sorted, importance_sorted):
  counter += 1
  print('\'' + imp + '\',')
  if counter == 100:
    break

# **Testing**:

In [ ]:
nu_list = [0.01, 0.1, 0.25, 0.5]
gamma_list = [0.1]
degree_list = [6, 4]
tol_list = [0.01, 0.001]
coef0_list = [0, 0.01]
shrinking_list = [True]


best_acc = 0
best_set = []

for v1 in nu_list:
    for v2 in gamma_list:
        for v4 in tol_list:
            for v5 in coef0_list:
                for v6 in shrinking_list:
                    for v3 in degree_list:
                        clf = OneClassSVM(gamma=v2, nu=v1, kernel='poly', degree=v3, tol=v4, coef0=v5, shrinking=v6).fit(train_X)

                        predictions = clf.predict(test_X)

                        acc = accuracy_score(df_test_Y, predictions)*100

                        if acc > best_acc:
                            best_acc = acc
                            best_set = [v1, v2, v3, v4, v5, v6]
                            print('\nnu: ', v1, 'gamma:', v2, 'degree:', v3, 'tol:', v4, 'coef0:', v5, 'shrinking:', v6)
                            print("Accuracy: ", acc, "%")
                            print("AUC: ", roc_auc_score(df_test_Y, predictions))
                            print("F1: ", f1_score(df_test_Y, predictions, average='macro'))
                            print("Recall: ", recall_score(df_test_Y, predictions, average='macro'))
                            print("Precision: ", precision_score(df_test_Y, predictions, average='macro'))
                            tn, fp, fn, tp = confusion_matrix(df_test_Y, predictions).ravel()
                            print("Confusion Matrix:\n", tn, fp, fn, tp)


nu:  0.01 gamma: 0.1 degree: 6 tol: 0.01 coef0: 0 shrinking: True
Accuracy:  83.88975155279503 %
AUC:  0.8388975155279503
F1:  0.8355598071522501
Recall:  0.8388975155279503
Precision:  0.8688437958215653
Confusion Matrix:
 1264 24 391 897

nu:  0.01 gamma: 0.1 degree: 4 tol: 0.01 coef0: 0 shrinking: True
Accuracy:  85.79192546583852 %
AUC:  0.8579192546583851
F1:  0.855939242474527
Recall:  0.8579192546583851
Precision:  0.8787413729317362
Confusion Matrix:
 1256 32 334 954

nu:  0.01 gamma: 0.1 degree: 4 tol: 0.01 coef0: 0.01 shrinking: True
Accuracy:  85.83074534161491 %
AUC:  0.858307453416149
F1:  0.8563460870030213
Recall:  0.858307453416149
Precision:  0.8790063780690831
Confusion Matrix:
 1256 32 333 955


In [37]:
clf = OneClassSVM(gamma=0.1, nu=0.01, kernel='poly', degree=4, tol=0.01, coef0=0.01, shrinking=True).fit(train_X)

predictions = clf.predict(test_X)

acc = accuracy_score(df_test_Y, predictions)*100
print("Accuracy: ", acc, "%")
#print("AUC: ", roc_auc_score(df_test_Y, predictions))
print("F1: ", f1_score(df_test_Y, predictions, average='macro'))
print("Recall: ", recall_score(df_test_Y, predictions, average='macro'))
print("Precision: ", precision_score(df_test_Y, predictions, average='macro'))
tn, fp, fn, tp = confusion_matrix(df_test_Y, predictions).ravel()
print("Confusion Matrix:\n", tn, fp, fn, tp)

Accuracy:  85.83074534161491 %
F1:  0.8563460870030213
Recall:  0.858307453416149
Precision:  0.8790063780690831
Confusion Matrix:
 1256 32 333 955


In [15]:
clf = OneClassSVM(gamma='auto', nu=0.7, kernel='rbf').fit(train_X)

predictions = clf.predict(test_X)

print("Accuracy: ", accuracy_score(df_test_Y, predictions)*100, "%")
print("AUC: ", roc_auc_score(df_test_Y, predictions))
print("F1: ", f1_score(df_test_Y, predictions, average='macro'))
print("Recall: ", recall_score(df_test_Y, predictions, average='macro'))
print("Precision: ", precision_score(df_test_Y, predictions, average='macro'))
print("Confusion Matrix:\n", confusion_matrix(df_test_Y, predictions))

Accuracy:  14.221419185282521 %
AUC:  0.14221419185282524
F1:  0.1418661859037348
Recall:  0.14221419185282524
Precision:  0.14163286575833522
Confusion Matrix:
 [[ 1858 13362]
 [12749  2471]]


In [31]:
clf = OneClassSVM(gamma='auto', nu=0.8, kernel='sigmoid').fit(train_X)

predictions = clf.predict(test_X)

print("Accuracy: ", accuracy_score(df_test_Y, predictions)*100, "%")
print("AUC: ", roc_auc_score(df_test_Y, predictions))
print("F1: ", f1_score(df_test_Y, predictions, average='macro'))
print("Recall: ", recall_score(df_test_Y, predictions, average='macro'))
print("Precision: ", precision_score(df_test_Y, predictions, average='macro'))
print("Confusion Matrix:\n", confusion_matrix(df_test_Y, predictions))

Accuracy:  37.877792378449406 %
AUC:  0.37877792378449404
F1:  0.3090172824258485
Recall:  0.37877792378449404
Precision:  0.29666376482525636
Confusion Matrix:
 [[  929 14291]
 [ 4619 10601]]
